### PR TITLE
V1 alpha1 k8s client

### PR DESCRIPTION
**Pull request by davidhadas**
This PR includes auto-generation of DeepCopy and clientset code for a Guardian 
Guardian maintains the micro-rules for every service protected by Guard. 
Each Guardian is composed of a few control flags (GuardianSpec.Control)
The learned micro rules (GuardianSpec.Learned)
The configured micro rules (GuardianSpec.Configured)

### CODE REVIEW

**Comment by davidhadas** [comment]
I intentionally left the actual code generated + the changes made to the vendor separated (not in this PR) - this results in a test failure. Let me know if you see a better way to do that.

**Review by maximilien — COMMENTED** [review]
Is this mostly API gen code?

**Comment by davidhadas** [comment]
This PR is what is needed to create the gen code. It does not include the gen code itself.

**Comment by davidhadas** [comment]
Please ignore the `verify / Verify Deps and Codegen` error and when ready /lgtm despite of it. It is complaining that we have not yet committed the generated code - which is planned to be sent as a separate PR to ease PR reviews. 

Tnx @rhuss for laying up a reasonable sequence for a code generation case. 
The sequence used:
1. Code generation
2. Generated code
3. Dependencies

**Comment by psschwei** [comment]
Think we'll need @evankanderson for the override

**Comment by evankanderson** [comment]
> Please ignore the `verify / Verify Deps and Codegen` error and when ready /lgtm despite of it. It is complaining that we have not yet committed the generated code - which is planned to be sent as a separate PR to ease PR reviews.

Does this mean that the repo won't build when this is committed until #38 is committed?

We generally add vendored code at the same time as the calls and updates `go.mod`. With #38 merged, maybe "verify deps and codegen" will work now?

In terms of review load, the `linguist-generated` annotation in `.gitattributes` ([sample](https://github.com/knative/serving/blob/main/.gitattributes)) can be used to cause github to auto-hide changes to generated files.

**Comment by evankanderson** [comment]
> Think we'll need @evankanderson for the override

It looks like the "Verify Deps and Codegen" GitHub action is not a blocking check on this repo, as you can see when inspecting the "tide" check, which is blocked on the `lgtm` label.

So we don't actually block the project from being in an unclean state at the moment, as we're not making that check be required by Prow. ("tide" is the prow component responsible for merging requests which are ready to merge)

**Comment by psschwei** [comment]
Whoops, I knew https://github.com/knative/test-infra/pull/3453/ made all GHAs required by default but didn't check the repo-specific settings

**Review by maximilien — APPROVED** [review]
Since this is generated code
/approve
/lgtm

**Comment by davidhadas** [comment]
> We generally add vendored code at the same time as the calls and updates `go.mod`. With #38 merged, maybe "verify deps and codegen" will work now?

This was tried after merging #38 and it still does not merge. 
It also doe snot merge after /lgtm

**Comment by davidhadas** [comment]
I have tried adding generated code, but it seems this will not do as well - PR #38 was reverted by #39 and #40, so we again do not have sufficient dependencies. The result is the `verify` fails - so it is the same with and without the dependencies... 

Bottom line, we need to override `verify` to make progress

**Comment by psschwei** [comment]
Saw last night that after the `lgtm` was added tide was holding on the verify job, so either we'll need @evankanderson 's superpowers or we should add the missing parts to this PR to make the verify job pass. I'd probably go with the second option...

**Comment by davidhadas** [comment]
ok, this means I will be changing this PR to include whatever we already have here + generated code + new dependencies  - it is going to be a very large PR  :)

**Comment by davidhadas** [comment]
image
This is after doing a tidy, vendor, update_codegen, tidy, vendor...

My k8s.io/hack are indeed 0x644
image

**Comment by davidhadas** [comment]
image

Scripts expect that you do not do go mod vendor before the commit..... not sure why this makes sense

**Comment by psschwei** [comment]
Not sure if this would fix the issue, but it's better to run `./hack/update-deps.sh` instead of `go mod vendor`

**Comment by davidhadas** [comment]
@psschwei , this issue with update_codegen was solved - can we redo the /lgtm here?

File: pkg/apis/wsecurity/v1alpha1/guardianSpec.go
Comment (psschwei) [inline pkg/apis/wsecurity/v1alpha1/guardianSpec.go:34]: Would `ActivateAll()` be a better name? (`Auto` feels a bit vague in terms of what the method is doing)

**Review by psschwei — COMMENTED** [review]
couple of minor things

File: pkg/apis/wsecurity/v1alpha1/guardianSpec.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1alpha1/guardianSpec.go:34]: ActivateAll is not a better name here :) since we are not activating blocking for example. 
It is a setup suitable for maximal automation and requires no user intervention. 
A good start for new services, while in development, and possibly when they start production and before you know what micro-rules to set. 

You can suggest a name :)

File: pkg/apis/wsecurity/register.go
Comment (davidhadas) [reply pkg/apis/wsecurity/register.go:23]: so, what should it be?

Comment (davidhadas) [reply pkg/apis/wsecurity/register.go:23]: changed to guard.security.knative.dev as discussed with @evankanderson

File: pkg/apis/wsecurity/v1alpha1/guardianSpec.go
Comment (davidhadas) [reply pkg/apis/wsecurity/v1alpha1/guardianSpec.go:34]: Changed to SetToMaximalAutomation()

Comment (psschwei) [reply pkg/apis/wsecurity/v1alpha1/guardianSpec.go]: works for me

**Review by psschwei — COMMENTED** [review]
/lgtm

**Comment by psschwei** [comment]
ugh... I just approved the deps update which is going to force a rebase :frowning_face:

**Review by psschwei — COMMENTED** [review]
/lgtm

**Pull request merged**
